File: trunk/uealib/src/readers/CoordsGraphReader.java
package readers;

import java.util.Scanner;

/**
 * Читает полный граф, заданный координатами вершин в двухмерном пространстве.
 * Первая строка файла должна содержать число вершин, последующие --
 * координаты вершин, разделенные пробелами.
 */
public class CoordsGraphReader extends GraphReader {
	

	protected Graph readGraph(Scanner s) {
		int n = s.nextInt();
		double[][] coords = new double[n][2];
		
		for ( int i = 0; i < n; ++i ) {
			double[] point = new double[2];
			point[0] = s.nextDouble();
			point[1] = s.nextDouble();
			coords[i] = point;
		}

		double[][] matrix = this.computeWeights(coords);

		Graph result = new Graph();
		result.setWeights(matrix);
		
		return result;
	}
	
	/**
	 * Вычисляет матрицу весов по списку вершин
	 */
	private double[][] computeWeights(double[][] coords) {
		int n = coords.length;
		double[][] matrix = new double[n][n];
		
		for ( int i = 0; i < n; ++i ) {
			for ( int j = 0; j < n; ++j ) {
				matrix[i][j] = distance(coords[i], coords[j]);
			}
		}
		
		return matrix;
	}
	
	/**
	 * Возвращает расстояние между двумя точками, округлённое до int
	 */
	private double distance(double[] a, double[] b) {
		return Math.sqrt( Math.pow(b[0] - a[0], 2) + Math.pow(b[1] - a[1], 2) );
	}

}

File: taboosearch/src/readers/CoordsGraphReader.scala
package core
import java.util.Scanner;

class CoordsGraphReader extends GraphReader {
	
	type Point = Pair[Double, Double]
	
	override def readGraph(s : Scanner) : Graph = {
		val n = s.nextInt();
		val coords = new Array[Point](n);
		
		for (i <- 0 until n) {
			val x = s.nextDouble();
			val y = s.nextDouble();
			coords(i) = new Point(x, y);
		}
		
		val weights = computeWeights(coords);		
		return new Graph(weights);
	}
	
	private def computeWeights(coords : Array[Point]) : Array[Array[Double]] = {
		val n = coords.length;
		val matrix = Array.ofDim[Double](n, n);
		for (i <- 0 until n) {
			for (j <- 0 until n) {
				val dist = distance(coords(i), coords(j));
				matrix(i)(j) = dist; 
				matrix(j)(i) = dist;
			}
		}
		return matrix;
	}
	
	private def distance(a : Point, b : Point) : Double = {
		return Math.sqrt(Math.pow(b._1 - a._1, 2) + Math.pow(b._2 - a._2, 2));
	}

}